libraries/AP_Quicktune/AP_Quicktune.cpp

```diff
@@ -0,0 +1,690 @@
+        if (reduction_pct > reduce_max) {
+            const float new_value = saved_value * (100 - reduce_max) * 0.01;
+            GCS_SEND_TEXT(MAV_SEVERITY_INFO, "Limiting %s %.3f -> %.3f", get_param_name(param), value, new_value);
+            value = new_value;
+        }
+    }
+   return value;
+}
+
+const char* AP_Quicktune::get_param_name(AP_Quicktune::Param param) const
+{
+    switch (param)
+    {
+        case Param::RLL_P:
+            return "Roll P";
+        case Param::RLL_I:
+            return "Roll I";
+        case Param::RLL_D:
+            return "Roll D";
+        case Param::PIT_P:
+            return "Pitch P";
+        case Param::PIT_I:
+            return "Pitch I";
+        case Param::PIT_D:
+            return "Pitch D";
+        case Param::YAW_P:
+            return "Yaw P";
+        case Param::YAW_I:
+            return "Yaw I";
+        case Param::YAW_D:
+            return "Yaw D";
+        default:
+            INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
+            return "UNK";
+    }
+}
+
+float AP_Quicktune::get_gain_mul() const
+{
+   return expf(logf(2.0)/(UPDATE_RATE_HZ*double_time));
+}
+
+void AP_Quicktune::restore_all_params()
+{
+    for (uint8_t p = 0; p < uint8_t(Param::END); p++) {
+        const auto param = Param(p);
+        if (BIT_IS_SET(param_changed, p)) {
+            set_param_value(param, param_saved[p]);
+            BIT_CLEAR(param_changed, p);
+        }
+    }
+}
+
+void AP_Quicktune::save_all_params()
+{
+    for (uint8_t p = 0; p < uint8_t(Param::END); p++) {
+        const auto param = Param(p);
+        set_and_save_param_value(param, get_param_value(param));
+        param_saved[p] = get_param_value(param);
+        BIT_CLEAR(param_changed, p);
+    }
+    GCS_SEND_TEXT(MAV_SEVERITY_NOTICE, "Quicktune: Saved");
+}
+
+AP_Quicktune::Param AP_Quicktune::get_pname(AP_Quicktune::AxisName axis, AP_Quicktune::Stage stage) const
+{
+    const uint8_t axis_offset = uint8_t(axis) * param_per_axis;
+    switch (stage)
+    {
+    case Stage::P:
+        return Param(uint8_t(Param::RLL_P) + axis_offset);
+    case Stage::D:
+        return Param(uint8_t(Param::RLL_D) + axis_offset);
+    case Stage::FLTT:
+        return Param(uint8_t(Param::RLL_FLTT) + axis_offset);
+    case Stage::FLTD:
+        return Param(uint8_t(Param::RLL_FLTD) + axis_offset);
+    default:
+        INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
+        return Param::END;
+    }
+}
+
+AP_Quicktune::Stage AP_Quicktune::get_stage(AP_Quicktune::Param param) const
+{
+    if (param == Param::RLL_P || param == Param::PIT_P || param == Param::YAW_P) {
+        return Stage::P;
+    } else {
+        return Stage::END; //Means "anything but P gain"
+    }
+}
+
+AP_Float *AP_Quicktune::get_param_pointer(AP_Quicktune::Param param) const
+{
+    const AxisName axis = AxisName(uint8_t(param) / param_per_axis);
```

Comment:
You could use `get_axis` here.
```suggestion
    const AxisName axis = get_axis(param);
```

Reply:
changed